cranelift/codegen/src/machinst/lower.rs

```diff
@@ -583,27 +611,48 @@ impl<'func, I: VCodeInst> Lower<'func, I> {
         // end of the BB (in the toplevel driver `lower()`).
         for inst in self.f.layout.block_insts(block).rev() {
             let data = &self.f.dfg[inst];
+            let has_side_effect = has_lowering_side_effect(self.f, inst);
+            // If side-effectful inst has been sunk to another location, skip
+            // it.
+            let inst_sunk = has_side_effect && self.inst_sunk.contains(&inst);
+            if inst_sunk {
+                continue;
+            }
+            // Are any outputs used at least once?
             let value_needed = self
                 .f
                 .dfg
                 .inst_results(inst)
                 .iter()
-                .any(|&result| self.vreg_needed[self.value_regs[result].get_index()]);
+                .any(|&result| self.value_lowered_uses[result] > 0);
```

Comment:
ditto for this bit. maybe we could have a `is_result_needed` helper

Reply:
Done.